sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -161,22 +164,51 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           bucketSpec,
           Some(partitionSchema))
 
+        val catalogTable = metastoreRelation.catalogTable
         val logicalRelation = cached.getOrElse {
           val sizeInBytes =
             metastoreRelation.stats(sparkSession.sessionState.conf).sizeInBytes.toLong
           val fileIndex = {
-            val index = new CatalogFileIndex(
-              sparkSession, metastoreRelation.catalogTable, sizeInBytes)
+            val index = new CatalogFileIndex(sparkSession, catalogTable, sizeInBytes)
             if (lazyPruningEnabled) {
               index
             } else {
               index.filterPartitions(Nil)  // materialize all the partitions in memory
             }
           }
           val partitionSchemaColumnNames = partitionSchema.map(_.name.toLowerCase).toSet
-          val dataSchema =
-            StructType(metastoreSchema
-              .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+          val filteredMetastoreSchema = StructType(metastoreSchema
+            .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
+
+          val inferenceMode = sparkSession.sessionState.conf.caseSensitiveInferenceMode
+          val shouldInfer = (inferenceMode != NEVER_INFER) && !catalogTable.schemaPreservesCase
+
+          val (dataSchema, updatedTable) = if (shouldInfer) {
+            logInfo(s"Inferring case-sensitive schema for table $tableName (inference mode: " +
+              s"$inferenceMode)")
+            val inferredSchema
+              = inferSchema(metastoreSchema, options, defaultSource, fileType, fileIndex)
+
+            // If configured, save the inferred case-sensitive schema to the table properties and
+            // fetch the updated CatalogTable record for use in the LogicalRelation.
+            val updatedTable = if (inferenceMode == INFER_AND_SAVE) {
+              updateCatalogTable(catalogTable, inferredSchema).getOrElse {
+                logWarning(s"Unable to save case-sensitive schema for table $tableName " +
+                  s"(inference mode: $inferenceMode)")
+                catalogTable
+              }
+            } else {
+              catalogTable
+            }
+            val schema = inferredSchema.getOrElse {
```

Comment:
Should we put this before `updatedTable` above? If the schema is unable to be inferred, we don't need to update the table.